File: py/src/util/generate_single_copy_gene.py
import sys

# python generate_single_copy_gene.py <gene_to_spe_list> <m8_file> <output_file>

gene_to_spe_list = sys.argv[1]
m8_file = sys.argv[2]
output_file = sys.argv[3]

gene_to_spe = {}

species_set = set()
# Read in the gene to species list
with open(gene_to_spe_list, 'r') as f:
    for line in f:
        line = line.strip()
        af_gene, spe = line.split()
        if af_gene[:3] == 'AF-':
            gene = af_gene.split('-')[1]
        else:
            gene = af_gene
        if gene not in gene_to_spe:
            gene_to_spe[gene] = set()
        gene_to_spe[gene].add(spe)
        species_set.add(spe)

species_count = len(species_set)

# Read in the m8 file
# and output the gene, multiple copy percent, and single copy percent

with open(m8_file, 'r') as f, open(output_file, 'w') as out:
    curr_query = None
    spe_dict = {}
    for line in f:
        line = line.strip()
        query, target = line.split()[:2]
        if target[:3] == 'AF-':
            target = target.split('-')[1]
        if query == curr_query:
            species = gene_to_spe[target]
            for spe in species:
                if spe not in spe_dict:
                    spe_dict[spe] = 0
                spe_dict[spe] += 1
        else:
            if curr_query is not None:
                single_copy = sum([1 for spe in spe_dict if spe_dict[spe] == 1])
                single_copy_percent = single_copy / species_count
                multiple_copy = len(spe_dict)
                multiple_copy_percent = multiple_copy / species_count
                out.write(curr_query + '\t' + str(multiple_copy_percent) + '\t' + str(single_copy_percent) + '\n')
            curr_query = query
            spe_dict = {}
            species = gene_to_spe[target]
            for spe in species:
                if spe not in spe_dict:
                    spe_dict[spe] = 0
                spe_dict[spe] += 1
    single_copy = sum([1 for spe in spe_dict if spe_dict[spe] == 1])
    single_copy_percent = single_copy / species_count
    multiple_copy = len(spe_dict)
    multiple_copy_percent = multiple_copy / species_count
    out.write(curr_query + '\t' + str(multiple_copy_percent) + '\t' + str(single_copy_percent) + '\n')

File: src/util/generate_single_copy_gene.rs
use std::collections::{HashMap, HashSet};
use std::env;
use std::fs::File;
use std::io::{self, BufRead, BufReader, BufWriter, Write};
use std::path::Path;

pub fn profile(m8_file: &str, mapping: &str, output_dir: &str, threshold: &f32, print_copiness: &bool) -> io::Result<()> {
    let mut gene_to_spe: HashMap<String, HashSet<String>> = HashMap::new();
    let mut species_set: HashSet<String> = HashSet::new();

    // Read the gene to species list
    let file = File::open(mapping)?;
    let reader = BufReader::new(file);
    for line in reader.lines().filter_map(|l| l.ok()) {
        let parts: Vec<&str> = line.split_whitespace().collect();
        let af_gene = parts[0];
        let spe = parts[1].to_string();

        let gene = if af_gene.starts_with("AF-") {
            af_gene.split('-').nth(1).unwrap_or(af_gene)
        } else {
            af_gene
        }.to_string();

        gene_to_spe.entry(gene).or_insert_with(HashSet::new).insert(spe.clone());
        species_set.insert(spe);
    }

    let species_count = species_set.len() as f32;

    // Process the m8 file and output the statistics
    let mut output = File::create(format!("{}/copiness.tsv",output_dir))?;
    // Write out the first line
    writeln!(output, "Query\tMultipleCopyPercent\tSingleCopyPercent")?;
    let file = File::open(m8_file)?;
    let reader = BufReader::new(file);
    let mut curr_query: Option<String> = None;
    let mut spe_cnt: HashMap<String, i32> = HashMap::new();
    let mut gene2spe: HashMap<String, HashSet<String>> = HashMap::new();

    for line in reader.lines().filter_map(|l| l.ok()) {
        let parts: Vec<&str> = line.split_whitespace().collect();
        let query = parts[0].to_string();
        let target = parts[1];

        let target_gene = if target.starts_with("AF-") {
            target.split('-').nth(1).unwrap_or(target)
        } else {
            target
        };

        if Some(&query) != curr_query.as_ref() {
            if let Some(q) = curr_query.take() {
                output_statistics_and_genes(&mut output, &q, &spe_cnt, &gene2spe, species_count, *threshold, output_dir)?;
            }
            curr_query = Some(query);
            spe_cnt.clear();
            gene2spe.clear();
        }

        if let Some(species) = gene_to_spe.get(target_gene) {
            for spe in species {
                *spe_cnt.entry(spe.to_string()).or_insert(0) += 1;
                gene2spe.entry(spe.to_string()).or_insert_with(HashSet::new).insert(target_gene.to_string());
            }
        }
    }

    if let Some(q) = curr_query {
        output_statistics_and_genes(&mut output, &q, &spe_cnt, &gene2spe, species_count, *threshold, output_dir)?;
    }

    Ok(())
}

fn output_statistics_and_genes<W: Write>(output: &mut W, query: &str, spe_cnt: &HashMap<String, i32>, gene2spe: &HashMap<String, HashSet<String>>, species_count: f32, threshold: f32, output_dir: &str) -> io::Result<()> {
    let single_copy = spe_cnt.values().filter(|&&count| count == 1).count() as f32;
    let multiple_copy = spe_cnt.len() as f32;

    let single_copy_percent = single_copy / species_count;
    let multiple_copy_percent = multiple_copy / species_count;

    // Write out to copiness.tsv
    writeln!(output, "{}\t{}\t{}", query, multiple_copy_percent, single_copy_percent)?;

    // Write out the gene list if it is considered as core gene
    if single_copy_percent >= threshold {
        let output_path = Path::new(output_dir).join(format!("{}.txt", query.split('-').nth(1).unwrap_or(query)));
        let mut output_file = BufWriter::new(File::create(output_path)?);

        for (spe, targets) in gene2spe {
            if targets.len() == 1 {
                let target = targets.iter().next().unwrap();
                writeln!(output_file, "{}\t{}", target, spe)?;
            }
        }
    }

    Ok(())
}
